### PR TITLE
[Applications] - service and component upgrade

### PR DESCRIPTION
**Pull request by LeraKornienko**

### CODE REVIEW

File: src/app/shell/personal-cabinet/applications/applications.component.html [hunk 1]
```diff
@@ -64,44 +64,48 @@
...
+          <ng-container *ngTemplateOutlet="ParentApplication; context:{$implicit: applicationCard$ | async }">
           </ng-container>
         </mat-tab>
       </mat-tab-group>
     </div>
 
+  <app-paginator  [totalEntities]="(applicationsCard$ | async).totalAmount" [currentPage]="currentPage" [itemsPerPage]="applicationsPerPage$ | async"
+    (pageChange)="onPageChange($event)" (itemsPerPageChange)="onItemsPerPageChange($event)">
+  </app-paginator>
+
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/applications.component.html:96]: I see you have added paginator only for parents, move it after the container at the very beginning for both roles

File: src/app/shell/personal-cabinet/applications/applications.component.html [hunk 2]
```diff
@@ -64,44 +64,48 @@
...
+  <app-application-card *ngFor="let application of (applicationsCard$ | async)?.entities" [application]="application" [userRole]="role"
     (approved)="onApprove($event)" (rejected)="onReject($event)" (leave)="onLeave($event)"
     (infoShow)="onInfoShow($event)" (infoHide)="onInfoHide()">
   </app-application-card>
 </ng-template>
 
 <ng-template #ProviderApplication let-applications>
   <ng-container *ngIf="!(isLoadingCabinet$ | async)">
-    <ng-container *ngIf="applications?.length else isEmptyList;">
+    <ng-container *ngIf="(applicationCard$ | async)?.length else isEmptyList;">
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/applications.component.html:108]: too much of async, it should be replaced with subscribtion

File: src/app/shell/personal-cabinet/applications/applications.component.ts [hunk 3]
```diff
@@ -27,21 +32,30 @@ import { ChildDeclination, WorkshopDeclination } from 'src/app/shared/enum/enumU
 })
 export class ApplicationsComponent extends CabinetDataComponent implements OnInit, AfterViewInit {
 
+  @Select(PaginatorState.applicationsPerPage)
+  directionsPerPage$: Observable<number>;
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/applications.component.ts:36]: applicationsPerPage$

**Review by litvinets — CHANGES_REQUESTED** [review]

File: src/app/shell/personal-cabinet/applications/applications.component.html [hunk 4]
```diff
@@ -1,4 +1,13 @@
-<div class="container" *ngIf="(role === Role.provider) then isProvider; else isParent"></div>
+<div *ngIf="applicationCards" class="container">
+  <ng-container class="container" *ngIf="(role === Role.provider) then isProvider; else isParent"></ng-container>
+  <app-paginator
+  [totalEntities]="applicationCards.totalAmount"
+  [currentPage]="currentPage"
+  [itemsPerPage]="applicationsPerPage$ | async"
+  (pageChange)="onPageChange($event)"
+  (itemsPerPageChange)="onItemsPerPageChange($event)">
+  </app-paginator>
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/applications.component.html:9]: let's hide paginator while loading

File: src/app/shell/personal-cabinet/applications/applications.component.ts [hunk 5]
```diff
@@ -58,18 +75,21 @@ export class ApplicationsComponent extends CabinetDataComponent implements OnIni
     private route: ActivatedRoute,) {
     super(store, matDialog, actions$);
   }
-  ngAfterViewInit(): void {
-    this.tabGroup.selectedIndex = this.tabIndex;
-  }
+
+  ngAfterViewInit(): void {}
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/applications.component.ts:79]: remove if if it is not needed anymore

File: src/app/shell/personal-cabinet/applications/applications.component.ts [hunk 6]
```diff
@@ -27,21 +32,33 @@ import { ChildDeclination, WorkshopDeclination } from 'src/app/shared/enum/enumU
 })
 export class ApplicationsComponent extends CabinetDataComponent implements OnInit, AfterViewInit {
 
+  @Select(PaginatorState.applicationsPerPage)
+  applicationsPerPage$: Observable<number>;
+  @Select(UserState.applications)
+  applicationCards$: Observable<ApplicationCards>;
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/applications.component.ts:38]: there is already select in cabinet-data-component, so you  do not need to copy past it

File: src/app/shell/personal-cabinet/parent/children/children.component.html [hunk 7]
```diff
@@ -20,7 +20,7 @@
   <ng-container>
     <div class="card-wrapper">
       <app-child-card class="card" *ngFor="let child of (childrenCards$ | async)?.entities" [child]="child"
-        (deleteChild)="onDelete($event)" [applications]="childApplications(applications$ |async, child)">
+        (deleteChild)="onDelete($event)" [applications]="childApplications(applicationCards$ | async, child)">
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/parent/children/children.component.html:23]: this functionality is broken, as I understood it should be entities 
image

**Review by litvinets — CHANGES_REQUESTED** [review]
parent workshops are broken :((( 
image

**Comment by litvinets** [comment]
I don't see any changes for displaying the status "isBlocked' in the application card, so I think you forget to add it. I suggest checking *ngIf="application.isBlocked" for this code and id the applications is blocked, then display blocked status and its reason 
image

**Review by Pabloostr — APPROVED** [review]

File: src/app/shell/personal-cabinet/applications/applications.component.ts [hunk 8]
```diff
@@ -142,13 +157,16 @@ export class ApplicationsComponent extends CabinetDataComponent implements OnIni
...
+    this.applicationParams.status = status;
+
     if (this.role === Role.provider) {
-      this.providerApplicationParams.status = status;
-      this.getProviderApplications(this.providerApplicationParams);
+      this.applicationParams.showBlocked = tabLabel === ApplicationTitlesReverse[ApplicationTitles.Blocked];
+      this.getProviderApplications(this.applicationParams);
     } else {
-      this.getParentApplications(status);
+      this.applicationParams.showBlocked = false;
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/applications.component.ts:168]: I belive it is false by default, so you can remove it

**Review by litvinets — CHANGES_REQUESTED** [review]

**Review by litvinets — CHANGES_REQUESTED** [review]

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts [hunk 9]
```diff
@@ -70,7 +71,10 @@ export class ApplicationCardComponent implements OnInit {
     this.subrole$
       .pipe(takeUntil(this.destroy$))
       .subscribe((subrole: string) => (this.subrole = subrole));
-
+    if (this.application.isBlocked) {
+      const applicationStatusBlocked = "Blocked";
+      this.application.status = applicationStatusBlocked;
+    }
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts:77]: remove

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts [hunk 10]
```diff
@@ -58,6 +58,7 @@ export class ApplicationCardComponent implements OnInit {
   @Output() leave = new EventEmitter();
   @Output() infoShow = new EventEmitter();
   @Output() infoHide = new EventEmitter();
+  @Output() passApplication = new EventEmitter();
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts:61]: remove

File: src/app/shell/personal-cabinet/applications/applications.component.ts [hunk 11]
```diff
@@ -57,38 +71,39 @@ export class ApplicationsComponent extends CabinetDataComponent implements OnIni
     private router: Router,
     private route: ActivatedRoute,) {
     super(store, matDialog, actions$);
+    this.route.queryParams
+      .pipe(takeUntil(this.destroy$))
+      .subscribe((params: Params) => this.tabIndex = Object.keys(ApplicationTitles).indexOf(params['status']));
+
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/applications.component.ts:77]: remove

File: src/app/shared/components/status-info-card/status-info-card.component.html [hunk 12]
```diff
@@ -1,12 +1,13 @@
...
-  <div class="application-info-menu application-info-description-card">({{
-    applicationStatusDescription[application.status] }})</div>
-    <p *ngIf="application.rejectionMessage" class="application-info-menu">{{application.rejectionMessage}}</p>
+    <div class="application-popup-status {{application.status}}">
+      <span> <i class={{applicationIcons[application.status]}}></i></span>
+      <span> {{applicationTitles[application.status]}}</span>
+    </div>
+    <div class="application-info-menu application-info-description-card">({{
+      applicationStatusDescription[application.status] }})</div>
+      <p *ngIf="application.rejectionMessage" class="application-info-menu">{{application.rejectionMessage}}</p>
```
Comment (litvinets) [inline src/app/shared/components/status-info-card/status-info-card.component.html:8]: in ts file     this.status = this.application.isBlocked ? ApplicationStatus.Blocked : ApplicationStatus[this.application.status];

File: src/app/shared/store/user.state.ts [hunk 13]
```diff
@@ -210,26 +212,24 @@ export class UserState {
...
     return this.applicationService
-      .getApplicationsByParentId(id, status)
+      .getApplicationsByParentId(id, parameters)
       .pipe(
-        tap((applications: Application[]) => {
-          return patchState({ applications: applications, isLoading: false });
-        }));
-  }
+        tap((applicationCards: ApplicationCards) =>
+        patchState(applicationCards ? { applicationCards: applicationCards, isLoading: false } : { filteredWorkshops: {totalAmount: 0, entities: []}, isLoading: false }),));
```
Comment (litvinets) [inline src/app/shared/store/user.state.ts:221]: applicationCards rename state value

File: src/app/shell/admin-tools/platform/directions/directions.component.ts [hunk 14]
```diff
@@ -80,6 +80,7 @@ export class DirectionsComponent implements OnInit, OnDestroy {
 
   onItemsPerPageChange(itemsPerPage: number): void {
     this.store.dispatch(new SetDirectionsPerPage(itemsPerPage));
+    this.store.dispatch(new GetFilteredDirections());
```
Comment (litvinets) [inline src/app/shell/admin-tools/platform/directions/directions.component.ts:83]: this.store.dispatch([ew SetDirectionsPerPage(itemsPerPage), new GetFilteredDirections()]);

File: src/app/shell/personal-cabinet/parent/children/children.component.ts [hunk 15]
```diff
@@ -69,6 +75,7 @@ export class ChildrenComponent extends CabinetDataComponent implements OnInit, O
   onPageChange(page: PaginationElement): void {
     this.currentPage = page;
     this.store.dispatch(new OnPageChangeChildrens(page));
+    this.store.dispatch( new GetUsersChildren());
   }
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/parent/children/children.component.ts:79]: call dispatch only once

File: src/app/shell/result/workshop-cards-list/workshop-cards-list.component.ts [hunk 16]
```diff
@@ -40,6 +41,7 @@ export class WorkshopCardsListComponent implements OnInit, OnDestroy {
   onPageChange(page: PaginationElement): void {
     this.currentPage = page;
     this.store.dispatch(new OnPageChangeWorkshops(page));
+    this.store.dispatch(new GetFilteredWorkshops());
   }
```
Comment (litvinets) [inline src/app/shell/result/workshop-cards-list/workshop-cards-list.component.ts:45]: same

File: src/app/shell/result/workshop-map-view-list/workshop-map-view-list.component.ts [hunk 17]
```diff
@@ -130,6 +131,7 @@ export class WorkshopMapViewListComponent implements OnInit, OnDestroy {
   onPageChange(page: PaginationElement): void {
     this.currentPage = page;
     this.store.dispatch(new OnPageChangeWorkshops(page));
+    this.store.dispatch(new GetFilteredWorkshops());
```
Comment (litvinets) [inline src/app/shell/result/workshop-map-view-list/workshop-map-view-list.component.ts:134]: same

**Review by litvinets — CHANGES_REQUESTED** [review]

**Review by litvinets — APPROVED** [review]
finally:)

**Review by perepichai — APPROVED** [review]

**Pull request merged**